loki/transformations/inline/procedures.py

```diff
@@ -87,12 +87,29 @@ def _map_unbound_dims(var, val):
         For example, mapping the passed array ``m(:,j)`` to the local
         expression ``a(i)`` yields ``m(i,j)``.
         """
+
+        def _offset_lbound(lbound, v):
+            return simplify(sym.Sum((sym.Sum((lbound, -1)), v)))
+
         new_dimensions = list(val.dimensions)
 
         indices = [index for index, dim in enumerate(val.dimensions) if isinstance(dim, sym.Range)]
 
         for index, dim in enumerate(var.dimensions):
-            new_dimensions[indices[index]] = dim
+            # if the argument contains an array range, we must map the bounds accordingly
+            if isinstance(val.dimensions[index], sym.Range) and (lower := val.dimensions[index].lower):
+                if isinstance(dim, sym.Range):
```

Comment:
[no action] Note to self, there's a similar index-shifting problem in the associate resolver. We could possibly re-use a shared utility for this if we were to externalise this? This is beyond this PR tho.